crates/typed-store/src/sally/mod.rs

```diff
@@ -0,0 +1,588 @@
+    where
+        J: Borrow<K>,
+    {
+        self.batch().delete_batch(self, keys)?.write()
+    }
+    fn try_catch_up_with_primary(&self) -> Result<(), Self::Error> {
+        match self {
+            SallyColumn::RocksDB((
+                db_map,
+                SallyConfig {
+                    mode: SallyRunMode::FallbackToDB,
+                },
+            )) => Ok(db_map.try_catch_up_with_primary()?),
+            SallyColumn::TestDB((
+                test_db,
+                SallyConfig {
+                    mode: SallyRunMode::FallbackToDB,
+                },
+            )) => Ok(test_db.try_catch_up_with_primary()?),
+        }
+    }
+}
+
+impl<J, K, U, V> TryExtend<(J, U)> for SallyColumn<K, V>
+where
+    J: Borrow<K> + std::clone::Clone,
+    U: Borrow<V> + std::clone::Clone,
+    K: Serialize,
+    V: Serialize,
+{
+    type Error = TypedStoreError;
+
+    fn try_extend<T>(&mut self, iter: &mut T) -> Result<(), Self::Error>
+    where
+        T: Iterator<Item = (J, U)>,
+    {
+        match self {
+            SallyColumn::RocksDB((
+                db_map,
+                SallyConfig {
+                    mode: SallyRunMode::FallbackToDB,
+                },
+            )) => db_map.try_extend(iter),
+            SallyColumn::TestDB((
+                test_db,
+                SallyConfig {
+                    mode: SallyRunMode::FallbackToDB,
+                },
+            )) => test_db.try_extend(iter),
+        }
+    }
+    fn try_extend_from_slice(&mut self, slice: &[(J, U)]) -> Result<(), Self::Error> {
+        match self {
+            SallyColumn::RocksDB((
+                db_map,
+                SallyConfig {
+                    mode: SallyRunMode::FallbackToDB,
+                },
+            )) => db_map.try_extend_from_slice(slice),
+            SallyColumn::TestDB((
+                test_db,
+                SallyConfig {
+                    mode: SallyRunMode::FallbackToDB,
+                },
+            )) => test_db.try_extend_from_slice(slice),
+        }
+    }
+}
+
+/// A Sally write batch provides a mutable struct which holds a collection of db mutation operations and
+/// applies them atomically to the db.
+pub enum SallyWriteBatch {
+    // Write batch for RocksDB backend when `fallback_to_db` is set as true
+    RocksDB(DBBatch),
+    // Write batch for btree map based backend
+    TestDB(TestDBWriteBatch),
+}
+
+impl SallyWriteBatch {
+    pub fn write(self) -> Result<(), TypedStoreError> {
+        match self {
+            SallyWriteBatch::RocksDB(db_batch) => db_batch.write(),
+            SallyWriteBatch::TestDB(write_batch) => write_batch.write(),
+        }
+    }
+    /// Deletes a set of keys given as an iterator
+    pub fn delete_batch<J: Borrow<K>, K: Serialize, V>(
+        self,
+        db: &SallyColumn<K, V>,
+        purged_vals: impl IntoIterator<Item = J>,
+    ) -> Result<Self, TypedStoreError> {
```

Comment:
does it make sense to keep the functional batch interface used by DBMap? As opposed to having a mutable batch which you can add new writes to. I've always found the existing interface to be a bit odd.

Reply:
Agree 100%, I do not like the current immutable interface - i'll make it mutable, which is more intuitive to use

Reply:
This is also done now (I had to add some new methods in DBMap<K, V> implementation to make this work but i will be deleting the original batch methods once we switch to sally)